src/loan.rs

```diff
@@ -1114,20 +1063,86 @@ impl Lender0 {
                     collateral_amount, tx_fee, collateral_input_amount,
                 )
             })?;
-        let (collateral_change_tx_out, _, _) = TxOut::new_last_confidential(
+
+        let mut tx_outs = vec![principal_tx_out];
+        let mut before_last_confidential_output_secrets = vec![TxOutSecrets::new(
+            self.usdt_asset_id,
+            abf_principal,
+            principal_amount.as_sat(),
+            vbf_principal,
+        )];
+
+        for (amount, address, asset) in [
+            (
+                collateral_change_amount,
+                borrower_address,
+                self.bitcoin_asset_id,
+            ),
+            (
+                principal_change_amount,
+                self.address.clone(),
+                self.usdt_asset_id,
+            ),
+        ]
+        .iter()
+        {
+            if amount == &Amount::ZERO {
+                log::info!("No change output needed for asset: {}", asset);
+                continue;
+            }
+
+            let (tx_out, abf, vbf) = TxOut::new_not_last_confidential(
+                rng,
+                secp,
+                amount.as_sat(),
+                address.clone(),
+                *asset,
+                &inputs_not_last_confidential,
+            )
+            .with_context(|| format!("could not construct change output for asset: {}", asset))?;
+
+            tx_outs.push(tx_out);
+            before_last_confidential_output_secrets.push(TxOutSecrets::new(
+                *asset,
+                abf,
+                amount.as_sat(),
+                vbf,
+            ))
+        }
```

Comment:
Personally, I find iterating over statically constructed lists confusing because they introduce state that you otherwise don't need to think about.

Trying to follow this code now requires you do think about how `before_last_confidential_output_secrets` and `tx_outs` are modified based on the contents of the list. If you were to unroll it, you could probably get away with constructing some of these lists absolutely (i.e. without needing mutation) which might be easier to understand.

Reply:
Fixed with: https://github.com/comit-network/baru/pull/58/commits/c34e56810c349e973b033d2030ccaafc2b22caca. I do agree that the original can be confusing.

Hopefully we can remove this sort of code by using PSETs.